plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts

```diff
@@ -0,0 +1,285 @@
+      },
+    );
+
+    const throttle = pThrottle({
+      limit: 10,
+      interval: durationToMilliseconds({ minutes: 1 }),
+    });
+    const throttled = throttle((opts: ChatPostMessageArguments) =>
+      this.sendNotification(opts),
+    );
+    this.sendNotifications = async (opts: ChatPostMessageArguments[]) => {
+      const results = await Promise.allSettled(
+        opts.map(message => throttled(message)),
+      );
+
+      let successCount = 0;
+      let failureCount = 0;
+
+      results.forEach(result => {
+        if (result.status === 'fulfilled') {
+          successCount++;
+        } else {
+          this.logger.error(
+            `Failed to send Slack channel notification: ${result.reason.message}`,
+          );
+          failureCount++;
+        }
+      });
+
+      this.messagesSent.add(successCount);
+      this.messagesFailed.add(failureCount);
+    };
+  }
+
+  getName(): string {
+    return 'SlackNotificationProcessor';
+  }
+
+  async processOptions(
+    options: NotificationSendOptions,
+  ): Promise<NotificationSendOptions> {
+    if (options.recipients.type !== 'entity') {
+      return options;
+    }
+
+    const entityRefs = [options.recipients.entityRef].flat();
+
+    const outbound: ChatPostMessageArguments[] = [];
+    await Promise.all(
+      entityRefs.map(async entityRef => {
+        const compoundEntityRef = parseEntityRef(entityRef);
+        // skip users as they are sent direct messages, but allow all other entity kinds
+        // to have a channel id annotation.
+        if (compoundEntityRef.kind === 'user') {
+          return;
+        }
+
+        let channel;
+        try {
+          channel = await this.getChannelId(entityRef);
+        } catch (error) {
+          this.logger.error(
+            `Failed to get Slack channel for entity: ${
+              (error as Error).message
+            }`,
+          );
+          return;
+        }
+
+        if (!channel) {
+          this.logger.debug(`No Slack channel found for entity: ${entityRef}`);
+          return;
+        }
+
+        this.logger.debug(
+          `Sending notification with payload: ${JSON.stringify(
+            options.payload,
+          )}`,
+        );
+
+        const payload = toChatPostMessageArgs({
+          channel,
+          payload: options.payload,
+        });
+
+        this.logger.debug(
+          `Sending Slack channel notification: ${JSON.stringify(payload)}`,
+        );
+        outbound.push(payload);
+      }),
+    );
+
+    console.log('dispatching message');
+    await this.sendNotifications(outbound);
+
+    return options;
+  }
+
+  async postProcess(
+    notification: Notification,
+    options: NotificationSendOptions,
+  ): Promise<void> {
+    if (options.recipients.type === 'broadcast' || !notification.user) {
+      return;
+    }
+
+    const entityRefs = [options.recipients.entityRef].flat();
+    if (entityRefs.some(e => parseEntityRef(e).kind === 'group')) {
```

Comment:
This actually cannot ever be group, it's always user. See https://github.com/backstage/backstage/blob/master/plugins/notifications-backend/src/service/router.ts#L535

I think there should be possibility to add some processor specific metadata to notification in `processOptions` which could in this case indicate that slack notification is already sent to group. This though requires some changes to the notification-backend 🤔

Reply:
Yes, the groups (and actually any other entities as well) are expanded before calling the `preProcess` and `postProcess` so basically this check is unnecessary. But AFAIK there's currently no way to pass processor specific data between the `processOptions`, `preProcess` and `postProcess` which could be used in this case. As said, this would require some additional changes to the `notifications-backend` itself.

Do you think you could live with sending both group and individual messages at this point or do you want to also make the necessary changes to `notifications-backend`? My proposal for this would be to add the following to `Notification` type:

```ts
metadata: JsonObject
```
and this would not be saved to the database so it would only live during the sending phase. WDYT?

Reply:
@drodil 
> This actually cannot ever be group, it's always user. See [master/plugins/notifications-backend/src/service/router.ts#L535](https://github.com/backstage/backstage/blob/master/plugins/notifications-backend/src/service/router.ts?rgh-link-date=2025-03-20T17%3A44%3A28Z#L535)

This is definitely how it works for us today -- notification.user is always a user, but options.recipients.entityRef **may** contain a group. This is how I've been determining whether or not we already sent a notification because the user comes from an expanded group.

I think we should aim to make the change you propose but we can do that separately. This iteration of the processor will only send a user a message if the notificationoptions also had a user as the recipient.